contracts/core/libs/utils/TagLib.sol

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.14;
+
+/*
+ * Each tag is encoded on 4 bits in the 256-bit bitmap.
+ * The 4 bits encode the tag's power (from 0 to 15) 
+ * If the power is 0, the tag is disabled. 
+
+     tag63               tag2         tag1         tag0
+   ┌────────────┐      ┌────────────┬────────────┬────────────┐
+   │ 1  1  1  1 │ .... │ 1  0  1  1 │ 0  0  0  0 │ 1  0  0  1 │
+   └────────────┘      └────────────┴────────────┴────────────┘
+      power 15            power 11     power 0      power 1
+    
+  * There is a total of 64 tags.
+  * The tag index must be between 0 and 63.
+  * The tag power must be between 0 and 15.
+  * An enabled tag will always have a power greater than 0.
+  * A disabled tag will always have a power equal to 0.
+ **/
+
+library TagLib {
+  uint256 constant MAX_INT = 2**256 - 1;
+
+  error TagIndexOutOfBounds(uint8 tagIndex);
+  error TagPowerOutOfBounds(uint8 power);
+
+  /**
+   * @dev Check if a tag, referenced by its tagIndex, is enabled in a uint256 number
+   * @param tagIndex index of the tag. Can be between 0 and 63
+   */
+  function _hasTag(uint256 self, uint8 tagIndex) internal pure returns (bool hashTag) {
+    uint256 currentTags = self;
+    uint8 tagPower = _getTagPower(currentTags, tagIndex);
+    // Check if the power of the tag is above 0, if yes the tag is enabled
+    return (tagPower) > 0;
+  }
+
+  /**
+   * @dev Return the power of tag, referenced by its tagIndex, of a uint256 number
+   * @param tagIndex index of the tag. Can be between 0 and 63
+   */
+  function _getTagPower(uint256 self, uint8 tagIndex) internal pure returns (uint8 tagPower) {
+    uint256 currentTags = self;
+    // Get the tag's 4 bits by shifting the tagIndex to the right
+    uint256 shifted = currentTags >> (4 * tagIndex);
+    // Get the tag power by masking the 4 bits encoding the power
+    return uint8(shifted & (2**4 - 1));
+  }
+
+  /**
+   * @dev Add a tag to a uint256 number
+   * @param tagIndex index of the tag. Can be between 0 and 63
+   * @param tagPower power of the tag. Can be between 0 and 15
+   */
+  function _addTag(
```

Comment:
Now like we don't have add and remove, i think we should rename this function and always call this one, even when the power is 0